README.md
# Silhouette

[![Twitter Follow](https://img.shields.io/twitter/follow/playframework?label=follow&style=flat&logo=twitter&color=brightgreen)](https://twitter.com/playframework)
[![Discord](https://img.shields.io/discord/931647755942776882?logo=discord&logoColor=white)](https://discord.gg/g5s2vtZ4Fa)
[![GitHub Discussions](https://img.shields.io/github/discussions/playframework/playframework?&logo=github&color=brightgreen)](https://github.com/playframework/playframework/discussions)
[![StackOverflow](https://img.shields.io/static/v1?label=stackoverflow&logo=stackoverflow&logoColor=fe7a16&color=brightgreen&message=playframework)](https://stackoverflow.com/tags/playframework)
[![YouTube](https://img.shields.io/youtube/channel/views/UCRp6QDm5SDjbIuisUpxV9cg?label=watch&logo=youtube&style=flat&color=brightgreen&logoColor=ff0000)](https://www.youtube.com/channel/UCRp6QDm5SDjbIuisUpxV9cg)
[![Twitch Status](https://img.shields.io/twitch/status/playframework?logo=twitch&logoColor=white&color=brightgreen&label=live%20stream)](https://www.twitch.tv/playframework)
[![OpenCollective](https://img.shields.io/opencollective/all/playframework?label=financial%20contributors&logo=open-collective)](https://opencollective.com/playframework)

[![Build Status](https://github.com/playframework/play-silhouette/actions/workflows/build-test.yml/badge.svg)](https://github.com/playframework/play-silhouette/actions/workflows/build-test.yml)
[![Maven](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette_2.13.svg?logo=apache-maven)](https://mvnrepository.com/artifact/org.playframework.silhouette/play-silhouette_2.13)
[![Repository size](https://img.shields.io/github/repo-size/playframework/play-silhouette.svg?logo=git)](https://github.com/playframework/play-silhouette)
[![Scala Steward badge](https://img.shields.io/badge/Scala_Steward-helping-blue.svg?style=flat&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA4AAAAQCAMAAAARSr4IAAAAVFBMVEUAAACHjojlOy5NWlrKzcYRKjGFjIbp293YycuLa3pYY2LSqql4f3pCUFTgSjNodYRmcXUsPD/NTTbjRS+2jomhgnzNc223cGvZS0HaSD0XLjbaSjElhIr+AAAAAXRSTlMAQObYZgAAAHlJREFUCNdNyosOwyAIhWHAQS1Vt7a77/3fcxxdmv0xwmckutAR1nkm4ggbyEcg/wWmlGLDAA3oL50xi6fk5ffZ3E2E3QfZDCcCN2YtbEWZt+Drc6u6rlqv7Uk0LdKqqr5rk2UCRXOk0vmQKGfc94nOJyQjouF9H/wCc9gECEYfONoAAAAASUVORK5CYII=)](https://scala-steward.org)
[![Mergify Status](https://img.shields.io/endpoint.svg?url=https://api.mergify.com/v1/badges/playframework/play-silhouette&style=flat)](https://mergify.com)

**Silhouette** is an authentication library for Play Framework applications that supports several authentication methods, including OAuth1, OAuth2, OpenID, CAS, Credentials, Basic Authentication, Two Factor Authentication or custom authentication schemes.

## Version compatibility matrix

|            | Play 3.x with Apache Pekko | Play 2.9.x with Akka | Play 2.8.x with Akka |
|------------|----------------------------|----------------------|----------------------|
| Silhouette | 10.x                       | 9.x                  | 8.x                  |

## Modules

| Module                              | Scala 3.x.x                                                                                                                                                                                                                                                                                                                  | Scala 2.13.x                                                                                                                                                                                                                                                                                                                                                | Scala 2.12.x                                                                                                                                                                                                                                                                                                                                                    |
|-------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **play-silhouette**                 | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette_3.svg?label=Maven%20Central%203&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette_3%22)           | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette_2.13.svg?label=Maven%20Central%202.13&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette_2.13%22)                                 | [![Maven Central](https://img.shields.io/maven-central/v/io.github.honeycomb-cheesecake/play-silhouette_2.12.svg?label=Maven%20Central%202.12&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22io.github.honeycomb-cheesecake%22%20AND%20a:%22play-silhouette_2.12%22)                                 |
| **play-silhouette-cas**             | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-cas_3.svg?label=Maven%20Central%203&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-cas_3%22)   | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-cas_2.13.svg?label=Maven%20Central%202.13&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-cas_2.13%22)                         | [![Maven Central](https://img.shields.io/maven-central/v/io.github.honeycomb-cheesecake/play-silhouette-cas_2.12.svg?label=Maven%20Central%202.12&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22io.github.honeycomb-cheesecake%22%20AND%20a:%22play-silhouette-cas_2.12%22)                         |
| **play-silhouette-crypto-jca**      | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-crypto-jca_3.svg?label=Maven%20Central%203&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-crypto-jca_3%22) | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-crypto-jca_2.13.svg?label=Maven%20Central%202.13&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-crypto-jca_2.13%22)           | [![Maven Central](https://img.shields.io/maven-central/v/io.github.honeycomb-cheesecake/play-silhouette-crypto-jca_2.12.svg?label=Maven%20Central%202.12&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22io.github.honeycomb-cheesecake%22%20AND%20a:%22play-silhouette-crypto-jca_2.12%22)           |
| **play-silhouette-password-argon2** | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-password-argon2_3.svg?label=Maven%20Central%203&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-password-argon2_3%22) | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-password-argon2_2.13.svg?label=Maven%20Central%202.13&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-password-argon2_2.13%22) | [![Maven Central](https://img.shields.io/maven-central/v/io.github.honeycomb-cheesecake/play-silhouette-password-argon2_2.12.svg?label=Maven%20Central%202.12&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22io.github.honeycomb-cheesecake%22%20AND%20a:%22play-silhouette-password-argon2_2.12%22) |
| **play-silhouette-password-bcrypt** | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-password-bcrypt_3.svg?label=Maven%20Central%203&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-password-bcrypt_3%22) | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-password-bcrypt_2.13.svg?label=Maven%20Central%202.13&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-password-bcrypt_2.13%22) | [![Maven Central](https://img.shields.io/maven-central/v/io.github.honeycomb-cheesecake/play-silhouette-password-bcrypt_2.12.svg?label=Maven%20Central%202.12&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22io.github.honeycomb-cheesecake%22%20AND%20a:%22play-silhouette-password-bcrypt_2.12%22) |
| **play-silhouette-persistence**     | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-persistence_3.svg?label=Maven%20Central%203&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-persistence_3%22) | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-persistence_2.13.svg?label=Maven%20Central%202.13&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-persistence_2.13%22)         | [![Maven Central](https://img.shields.io/maven-central/v/io.github.honeycomb-cheesecake/play-silhouette-persistence_2.12.svg?label=Maven%20Central%202.12&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22io.github.honeycomb-cheesecake%22%20AND%20a:%22play-silhouette-persistence_2.12%22)         |
| **play-silhouette-totp**            | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-totp_3.svg?label=Maven%20Central%203&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-totp_3%22) | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-totp_2.13.svg?label=Maven%20Central%202.13&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-totp_2.13%22)                       | [![Maven Central](https://img.shields.io/maven-central/v/io.github.honeycomb-cheesecake/play-silhouette-totp_2.12.svg?label=Maven%20Central%202.12&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22io.github.honeycomb-cheesecake%22%20AND%20a:%22play-silhouette-totp_2.12%22)                       |
| **play-silhouette-testkit**         | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-testkit_3.svg?label=Maven%20Central%203&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-testkit_3%22) | [![Maven Central](https://img.shields.io/maven-central/v/org.playframework.silhouette/play-silhouette-testkit_2.13.svg?label=Maven%20Central%202.13&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22org.playframework.silhouette%22%20AND%20a:%22play-silhouette-testkit_2.13%22)                 | [![Maven Central](https://img.shields.io/maven-central/v/io.github.honeycomb-cheesecake/play-silhouette-testkit_2.12.svg?label=Maven%20Central%202.12&style=for-the-badge&logo=scala&color=brightgreen&logoColor=green)](https://search.maven.org/search?q=g:%22io.github.honeycomb-cheesecake%22%20AND%20a:%22play-silhouette-testkit_2.12%22)                 |

## Installation

### Latest Production Versions

To get the latest production release(s) from this repository, add the following to your project's `build.sbt` file, replacing `x.x.x` with the `play-silhouette` version of choice:

```
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette" % "x.x.x"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-cas" % "x.x.x"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-crypto-jca" % "x.x.x"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-password-argon2" % "x.x.x"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-password-bcrypt" % "x.x.x"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-persistence" % "x.x.x"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-totp" % "x.x.x"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-testkit" % "x.x.x" % Test
```

### Latest Snapshot Versions

This fork of `play-silhouette` shall also be releasing SNAPSHOTS of the latest passed builds which can be used for the latest, bleeding-edge features, patches and code fixes prior to official production releases. If you want to pull these for testing with your code in development environments, please ensure that you add the appropriate resolver below, and select the appropriate version with `x.x.x` (>= 7.0.1).

```
resolvers += "snapshots" at "https://s01.oss.sonatype.org/content/repositories/snapshots/"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette" % "x.x.x-SNAPSHOT"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-cas" % "x.x.x-SNAPSHOT"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-crypto-jca" % "x.x.x-SNAPSHOT"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-password-argon2" % "x.x.x-SNAPSHOT"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-password-bcrypt" % "x.x.x-SNAPSHOT"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-persistence" % "x.x.x-SNAPSHOT"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-totp" % "x.x.x-SNAPSHOT"
libraryDependencies += "org.playframework.silhouette" %% "play-silhouette-testkit" % "x.x.x-SNAPSHOT" % Test
```

## Support

If you have question regarding Silhouette, please use the [chat]. **Please do not use the issue tracker for questions!**

## Contribution

Please read the [contributing guide] before you contribute. It contains very useful tips for a successful contribution.

## License

The code is licensed under [Apache License v2.0] and the documentation under [CC BY 3.0].

[GitHub]: https://github.com/playframework/play-silhouette
[GitHub issue tracker]: https://github.com/playframework/play-silhouette/issues
[chat]: https://discord.gg/g5s2vtZ4Fa
[Scala Style Guide]: http://docs.scala-lang.org/style/
[Apache License v2.0]: http://www.apache.org/licenses/LICENSE-2.0

**This repository started life as a fork from the [main repository](https://github.com/mohiva/play-silhouette) but now exists as it's own project. Thank you very much to previous maintainer Christian Kaps ([@akkie](https://github.com/akkie)) and all contributors for the work you've done to bring us to this point. This library was made fantastic and we are looking to continue along the path you set for us.**


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [10.0.0] - 2023-12-10

### Changed

- Support for Play 3.0.x with Apache Pekko thanks to [@MathisGuillet1](https://github.com/MathisGuillet1) and [@mkurz](https://github.com/mkurz)

## [9.0.0] - 2023-12-10

Play Silhouette is now under the Playframework organisation.

### Added

- Scala 3, Play 2.9.x compatibility thanks to [@ahjohannessen](https://github.com/ahjohannessen) and [@MathisGuillet1](https://github.com/MathisGuillet1)
- Commit information to discord bot snapshot message from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Commit and version information to discord bot production release message from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).

### Changed

- ***Package names has been changed from `io.github.honeycomb-cheesecake` to `org.playframework.silhouette`***
- Version to v9.0.0-SNAPSHOT for breaking change release from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- JodaTime replaced with Java Date API from [Rocketeer007](https://github.com/Rocketeer007).
- JWT provider from `com.atlassian.jwt.jwt-core` and `com.atlassian.jwt.jwt-api` to `com.auth0.java-jwt` to permit the use of signing algorithms other than HS256 from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Disable build failure on dependencies not being up-to-date

### Removed

- JodaTime dependency from [Rocketeer007](https://github.com/Rocketeer007).

## [8.0.2] - 2022-06-11

### Fixed

- sbt-sonatype configuration so that ```sonatypeCredentialHost``` points to correct location (```s01.oss.sonatype.org```) due to legacy host (```oss.sonatype.org```) is used by default which failed production release from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- sbt-sonatype configuration so that production ```publishTo``` points to local staging repository from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- scala.yml release pipeline so that ```+publishSigned``` is run prior to ```+sonatypeBundleRelease``` from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Corrected v8.0.1 by adding version information from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Corrected v8.0.0 by correcting date from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).

### Changed

- Updated scala.yml pipeline dynamic sonatype.sbt file to use environment variables rather than file from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).

## [8.0.1] - 2022-06-11

## Security

- Upgrading Play framework version to 2.8.16  to prevent [CVE-2022-31023](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2022-31023) from [@ndeverge](https://github.com/ndeverge).

### Fixed

- deploy-production-artifact GitHub Action workflow from **startsWith(github.ref, 'ref/tags/v')** to **startsWith(github.ref, 'refs/tags/v')** from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).

### Added

- Discord messages for pipeline runs from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).

### Changed

- Snapshot releases not executed on tagged artifacts as the sbt-sonatype plugin will just attempt to release a production version from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).

## [8.0.0] - 2022-05-16

### Fixed

- Code warnings from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Tranche of scapegoat results from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).

### Added

- Sbt release plugin to facilitate project releases from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Scapegoat stage to release process and github actions from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Dependency check stage to release process and github actions from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Dependency updates stage where build will fail with updates to release process and github actions from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Coveralls integration back into github actions from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Publishing step to github action for CI from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Create GitHub release step for tagged versions from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Publishing environment variables for CI deployment from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Import of GPG key step to github action for CI from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Disabled line for cross-build with scala 3 when supported by play from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Coveralls badge to README.md from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Discord badge to README.md from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).

### Changed

- Updated dependencies to latest from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Project build definitions from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Package definition from **com.mohiva.silhouette.play** to **io.github.honeycombcheesecake.play** from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Split github actions into build and test phases from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).
- Pull request template so it points to the honeycomb-cheesecake repository from [@honeycomb-cheesecake](https://github.com/honeycomb-cheesecake).

### Removed


project/BuildSettings.scala
/**
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import com.typesafe.sbt.SbtGhPages.GhPagesKeys._
import com.typesafe.sbt.SbtGhPages.ghpages
import com.typesafe.sbt.SbtGit.git
import com.typesafe.sbt.SbtSite.SiteKeys._
import com.typesafe.sbt.SbtSite.site
import sbt.Keys._
import sbt.{Def, _}
import sbtunidoc.ScalaUnidocPlugin.autoImport._


////*******************************
//// Basic settings
////*******************************
object BasicSettings extends AutoPlugin {
  override def trigger = allRequirements

  val `scalacOptions2.12.x` = Seq(
    "-Xlint:adapted-args",
    "-Ywarn-inaccessible",
    "-Ywarn-infer-any",
    "-Xlint:nullary-override",
    "-Xlint:nullary-unit",
    "-Xmax-classfile-name", "254",
    "-language:higherKinds"
  )

  val `scalacOptions2.13.x` = Seq(
    "-Xlint:adapted-args",
    "-Xlint:inaccessible",
    "-Xlint:infer-any",
    "-Xlint:nullary-override",
    "-Xlint:nullary-unit"
  )

  val scalacOptionsCommon = Seq(
    "-unchecked",
    "-deprecation",
    "-feature",
    "-encoding", "utf8",
    "-Xfatal-warnings",
    "-Xlint"
  )

  override def projectSettings = Seq(
    organization := "io.github.honeycomb-cheesecake",
    version := "7.0.8-SNAPSHOT",
    resolvers ++= Dependencies.resolvers,
    scalaVersion := "2.13.1",
    crossScalaVersions := Seq("2.13.1", "2.12.10"),
    crossVersion := CrossVersion.full,
    scalacOptions ++= {
      scalacOptionsCommon ++ (scalaBinaryVersion.value match {
        case "2.12" => `scalacOptions2.12.x`
        case "2.13" => `scalacOptions2.13.x`
      })
    },
    scalacOptions in Test ~= { options: Seq[String] =>
      options filterNot (_ == "-Ywarn-dead-code") // Allow dead code in tests (to support using mockito).
    },
    parallelExecution in Test := false,
    fork in Test := true,
    // Needed to avoid https://github.com/travis-ci/travis-ci/issues/3775 in forked tests
    // in Travis with `sudo: false`.
    // See https://github.com/sbt/sbt/issues/653
    // and https://github.com/travis-ci/travis-ci/issues/3775
    javaOptions += "-Xmx1G"
  )
}

////*******************************
//// Scalariform settings
////*******************************
object CodeFormatter extends AutoPlugin {

  import com.typesafe.sbt.SbtScalariform._

  import scalariform.formatter.preferences._

  lazy val BuildConfig = config("build") extend Compile
  lazy val BuildSbtConfig = config("buildsbt") extend Compile

  lazy val prefs = Seq(
    ScalariformKeys.preferences := ScalariformKeys.preferences.value
      .setPreference(FormatXml, false)
      .setPreference(DoubleIndentClassDeclaration, false)
      .setPreference(AlignSingleLineCaseStatements, true)
      .setPreference(DanglingCloseParenthesis, Preserve)
  )

  override def trigger = allRequirements

  override def projectSettings: Seq[Setting[_]] = defaultScalariformSettings ++ prefs ++
    inConfig(BuildConfig)(configScalariformSettings) ++
    inConfig(BuildSbtConfig)(configScalariformSettings) ++
    Seq(
      scalaSource in BuildConfig := baseDirectory.value / "project",
      scalaSource in BuildSbtConfig := baseDirectory.value / "project",
      includeFilter in (BuildConfig, ScalariformKeys.format) := ("*.scala": FileFilter),
      includeFilter in (BuildSbtConfig, ScalariformKeys.format) := ("*.sbt": FileFilter),
      ScalariformKeys.format in Compile := {
        (ScalariformKeys.format in BuildSbtConfig).value
        (ScalariformKeys.format in BuildConfig).value
        (ScalariformKeys.format in Compile).value
      }
    )
}

////*******************************
//// ScalaDoc settings
////*******************************
object Doc extends AutoPlugin {

  import play.core.PlayVersion

  override def projectSettings = Seq(
    autoAPIMappings := true,
    apiURL := Some(url(s"http://api.silhouette.mohiva.com/${version.value}/")),
    apiMappings ++= {
      implicit val cp: Classpath = (fullClasspath in Compile).value
      Map(
        jarFor("com.typesafe.play", "play") -> url(s"http://www.playframework.com/documentation/${PlayVersion.current}/api/scala/"),
        scalaInstance.value.libraryJar -> url(s"http://www.scala-lang.org/api/${scalaVersion.value}/")
      )
    }
  )

  /**
   * Gets the JAR file for a package.
   *
   * @param organization The organization name.
   * @param name The name of the package.
   * @param cp The class path.
   * @return The file which points to the JAR.
   * @see http://stackoverflow.com/a/20919304/2153190
   */
  private def jarFor(organization: String, name: String)(implicit cp: Seq[Attributed[File]]): File = {
    (for {
      entry <- cp
      module <- entry.get(moduleID.key)
      if module.organization == organization
      if module.name.startsWith(name)
      jarFile = entry.data
    } yield jarFile).head
  }
}

////*******************************
//// APIDoc settings
////*******************************
// @see https://github.com/paypal/horizon/blob/develop/src/main/scala/com/paypal/horizon/BuildUtilities.scala
object APIDoc {

  lazy val files = Seq(file("CNAME"))

  lazy val settings: Seq[Setting[_]] =
    site.settings ++
    ghpages.settings ++
    Seq(
      // Create version
      siteMappings ++= {
        val mapping = (mappings in (ScalaUnidoc, packageDoc)).value
        val ver = version.value
        for ((file, path) <- mapping) yield (file, s"$ver/$path")
      },
      // Add custom files from site directory
      siteMappings ++= baseDirectory.map { dir =>
        for (file <- files) yield (new File(dir.getAbsolutePath + "/site/" + file), file.name)
      }.value,
      // Do not delete old versions
      synchLocal := {
        val betterMappings = privateMappings.value.map { case (file, tgt) => (file, updatedRepository.value / tgt) }
        IO.copy(betterMappings)
        updatedRepository.value
      },
      git.remoteRepo := "git@github.com:honeycomb-cheesecake/play-silhouette.git"
    )
}

////*******************************
//// Maven settings
////*******************************
object Publish extends AutoPlugin {

  import xerial.sbt.Sonatype._

  override def trigger = allRequirements

  private val pom = {
    <scm>
      <url>git@github.com:honeycomb-cheesecake/play-silhouette.git</url>
      <connection>scm:git:git@github.com:honeycomb-cheesecake/play-silhouette.git</connection>
    </scm>
    <developers>
      <developer>
        <id>honeycomb-cheesecake</id>
        <name>Simon Ramzi</name>
        <url>https://github.com/honeycomb-cheesecake</url>
      </developer>
      <developer>
        <id>ndeverge</id>
        <name>Nicolas Deverge</name>
        <url>https://github.com/ndeverge</url>
      </developer>
    </developers>
  }

  override def projectSettings: Seq[Def.Setting[_]] = sonatypeSettings ++ Seq(
    description := "Authentication library for Play Framework applications that supports several authentication methods, including OAuth1, OAuth2, OpenID, CAS, Credentials, Basic Authentication, Two Factor Authentication or custom authentication schemes",
    homepage := Some(url("https://silhouette.readme.io/")),
    licenses := Seq("Apache License" -> url("https://github.com/honeycomb-cheesecake/play-silhouette/blob/master/LICENSE")),
    publishMavenStyle := true,
    publishArtifact in Test := false,
    pomIncludeRepository := { _ => false },
    pomExtra := pom
  )
}

////*******************************
//// Helpers
////*******************************
object Util {
  def priorTo213(scalaVersion: String): Boolean =
    CrossVersion.partialVersion(scalaVersion) match {
      case Some((2, minor)) if minor < 13 => true
      case _                              => false
    }
}

